sources/web/datalab/jupyter.ts

```diff
@@ -47,6 +47,25 @@ var jupyterServers: common.Map<JupyterServer> = {};
 var nextJupyterPort = 9000;
 
 /**
+ * Get the next available port and pass it to the given `resolved` callback.
+ */
+function getNextJupyterPort(resolved: (port: number)=>void, failed: (error: Error)=>void) {
+   var port = nextJupyterPort;
+   nextJupyterPort++;
+
+   tcp.check(port, '127.0.0.1').then(
```

Comment:
Can we put an upper limit on the number of ports we tried? For example from 9000~9200? There might be cases that the tcp.check call fails on any ports given 127.0.0.1, and we will end up in a dead loop.